File: simplified-books-core-ng/src/main/java/org/nypl/simplified/books/book_registry/BookRegistry.java
package org.nypl.simplified.books.book_registry;

import com.io7m.jfunctional.Option;
import com.io7m.jfunctional.OptionType;
import com.io7m.jnull.NullCheck;

import org.nypl.simplified.books.book_database.BookID;
import org.nypl.simplified.books.logging.LogUtilities;
import org.nypl.simplified.observable.Observable;
import org.nypl.simplified.observable.ObservableReadableType;
import org.nypl.simplified.observable.ObservableType;
import org.slf4j.Logger;

import java.util.Collections;
import java.util.HashSet;
import java.util.SortedMap;
import java.util.concurrent.ConcurrentSkipListMap;

import static org.nypl.simplified.books.book_registry.BookStatusEvent.Type.BOOK_CHANGED;
import static org.nypl.simplified.books.book_registry.BookStatusEvent.Type.BOOK_REMOVED;

public final class BookRegistry implements BookRegistryType {

  private static final Logger LOG = LogUtilities.getLog(BookRegistry.class);

  private final SortedMap<BookID, BookWithStatus> books_read_only;
  private final ConcurrentSkipListMap<BookID, BookWithStatus> books;
  private final ObservableType<BookStatusEvent> observable;

  private BookRegistry(
      final ConcurrentSkipListMap<BookID, BookWithStatus> books) {

    this.books = NullCheck.notNull(books, "Books");
    this.books_read_only = Collections.unmodifiableSortedMap(books);
    this.observable = Observable.create();
  }

  public static BookRegistryType create() {
    return new BookRegistry(new ConcurrentSkipListMap<>());
  }

  @Override
  public SortedMap<BookID, BookWithStatus> books() {
    return this.books_read_only;
  }

  @Override
  public ObservableReadableType<BookStatusEvent> bookEvents() {
    return this.observable;
  }

  @Override
  public OptionType<BookStatusType> bookStatus(final BookID id) {
    return book(id).map(BookWithStatus::status);
  }

  @Override
  public OptionType<BookWithStatus> book(final BookID id) {
    return Option.of(this.books.get(NullCheck.notNull(id, "id")));
  }

  @Override
  public void update(
      final BookWithStatus update) {

    NullCheck.notNull(update, "Update");
    this.books.put(update.book().id(), update);
    this.observable.send(BookStatusEvent.create(update.book().id(), BOOK_CHANGED));
  }

  @Override
  public void updateIfStatusIsMoreImportant(
      final BookWithStatus update) {

    NullCheck.notNull(update, "Update");

    final BookWithStatus current = this.books.get(update.book().id());
    if (current != null) {
      final BookStatusPriorityOrdering current_p = current.status().getPriority();
      final BookStatusPriorityOrdering update_p = update.status().getPriority();

      if (current_p.getPriority() <= update_p.getPriority()) {
        LOG.debug("current {} <= {}, updating", current, update);
        this.update(update);
        return;
      }

      LOG.debug("current {} > {}, not updating", current, update);
    } else {
      this.update(update);
    }
  }

  @Override
  public void clear() {
    final HashSet<BookID> ids = new HashSet<>(books.keySet());
    this.books.clear();
    for (final BookID id : ids) {
      this.observable.send(BookStatusEvent.create(id, BOOK_REMOVED));
    }
  }

  @Override
  public void clearFor(
      final BookID id) {
    this.books.remove(NullCheck.notNull(id, "ID"));
    this.observable.send(BookStatusEvent.create(id, BOOK_REMOVED));
  }
}

File: simplified-books-registry-api/src/main/java/org/nypl/simplified/books/book_registry/BookRegistry.kt
package org.nypl.simplified.books.book_registry

import com.io7m.jfunctional.FunctionType
import com.io7m.jfunctional.Option
import com.io7m.jfunctional.OptionType
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject
import org.nypl.simplified.books.api.BookID
import org.slf4j.LoggerFactory
import java.util.Collections
import java.util.SortedMap
import java.util.concurrent.ConcurrentSkipListMap

class BookRegistry private constructor(
  private val books: ConcurrentSkipListMap<BookID, BookWithStatus>
) : BookRegistryType {

  private val logger =
    LoggerFactory.getLogger(BookRegistry::class.java)
  private val booksReadOnly: SortedMap<BookID, BookWithStatus> =
    Collections.unmodifiableSortedMap(this.books)
  private val observable: PublishSubject<BookStatusEvent> =
    PublishSubject.create()

  override fun books(): SortedMap<BookID, BookWithStatus> {
    return this.booksReadOnly
  }

  override fun bookEvents(): Observable<BookStatusEvent> {
    return this.observable
  }

  override fun bookStatus(id: BookID): OptionType<BookStatus> {
    return this.book(id).map(FunctionType<BookWithStatus, BookStatus>(BookWithStatus::status))
  }

  override fun book(id: BookID): OptionType<BookWithStatus> {
    return Option.of(this.books[id])
  }

  override fun update(status: BookWithStatus) {
    val oldStatus = this.books[status.book.id]
    this.books[status.book.id] = status
    this.publishUpdateEvent(oldStatus, status)
  }

  private fun publishUpdateEvent(oldStatus: BookWithStatus?, newStatus: BookWithStatus) {
    if (newStatus.status == oldStatus?.status) {
      return
    }

    val event =
      if (oldStatus == null) {
        BookStatusEvent.BookStatusEventAdded(
          bookId = newStatus.book.id,
          statusNow = newStatus.status
        )
      } else {
        BookStatusEvent.BookStatusEventChanged(
          bookId = oldStatus.book.id,
          statusPrevious = oldStatus.status,
          statusNow = newStatus.status
        )
      }

    this.observable.onNext(event)
  }

  override fun updateIfStatusIsMoreImportant(status: BookWithStatus) {
    val current = this.books[status.book.id]
    if (current != null) {
      val currentPri = current.status.priority
      val updatePri = status.status.priority

      if (currentPri.priority <= updatePri.priority) {
        this.logger.debug("current {} <= {}, updating", current, status)
        this.update(status)
        return
      }

      this.logger.debug("current {} > {}, not updating", current, status)
    } else {
      this.update(status)
    }
  }

  override fun clear() {
    val entries = this.books.toMap()
    for (entry in entries) {
      this.observable.onNext(BookStatusEvent.BookStatusEventRemoved(entry.key, entry.value.status))
    }
    this.books.clear()
  }

  override fun clearFor(id: BookID) {
    val oldStatus = this.books.remove(id)
    if (oldStatus != null) {
      this.observable.onNext(BookStatusEvent.BookStatusEventRemoved(id, oldStatus.status))
    }
  }

  companion object {
    fun create(): BookRegistryType {
      return BookRegistry(ConcurrentSkipListMap())
    }
  }
}
